src/Microsoft.ApplicationInsights.AspNetCore/Implementation/TelemetryConfigurationOptionsSetup.cs

```diff
@@ -0,0 +1,102 @@
+namespace Microsoft.Extensions.DependencyInjection
+{
+    using System;
+    using System.Collections.Generic;
+    using Microsoft.ApplicationInsights.AspNetCore.Extensions;
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.Extensibility;
+#if NET451
+    using Microsoft.ApplicationInsights.Extensibility.PerfCounterCollector.QuickPulse;
+    using Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel;
+#endif
+    using Microsoft.Extensions.Options;
+
+    internal class TelemetryConfigurationOptionsSetup : IConfigureOptions<TelemetryConfiguration>
+    {
+        private readonly ApplicationInsightsServiceOptions applicationInsightsServiceOptions;
+        private readonly IEnumerable<ITelemetryInitializer> initializers;
+        private readonly IEnumerable<ITelemetryModule> modules;
+        private readonly ITelemetryChannel telemetryChannel;
+
+        public TelemetryConfigurationOptionsSetup(
+            IServiceProvider serviceProvider,
+            IOptions<ApplicationInsightsServiceOptions> applicationInsightsServiceOptions,
+            IEnumerable<ITelemetryInitializer> initializers,
+            IEnumerable<ITelemetryModule> modules)
+        {
+            this.applicationInsightsServiceOptions = applicationInsightsServiceOptions.Value;
+            this.initializers = initializers;
+            this.modules = modules;
+            this.telemetryChannel = serviceProvider.GetService<ITelemetryChannel>();
+        }
+
+        public void Configure(TelemetryConfiguration options)
+        {
+            if (this.applicationInsightsServiceOptions.InstrumentationKey != null)
+            {
+                options.InstrumentationKey = this.applicationInsightsServiceOptions.InstrumentationKey;
+            }
+
+            this.AddTelemetryChannelAndProcessorsForFullFramework(options);
+
+            options.TelemetryChannel = this.telemetryChannel ?? options.TelemetryChannel;
+
+            if (this.applicationInsightsServiceOptions.DeveloperMode != null)
+            {
+                options.TelemetryChannel.DeveloperMode = this.applicationInsightsServiceOptions.DeveloperMode;
+            }
+
+            if (this.applicationInsightsServiceOptions.EndpointAddress != null)
+            {
+                options.TelemetryChannel.EndpointAddress = this.applicationInsightsServiceOptions.EndpointAddress;
+            }
+
+            foreach (var initializer in this.initializers)
+            {
+                options.TelemetryInitializers.Add(initializer);
+            }
+
+            foreach (var module in this.modules)
```

Comment:
nitpick: There is no way to tell the type here or above, I don't know what var module or var initializer are when reading the source text in review unless I spend extra time scrolling back up to look for a definition like IEnumerable\<ITelemetryInitializer\>.  I know the "var" debate is a hot topic but my point is to increase readability, especially considering the time pressure we are all under to get this in.